src/data.rs

```diff
@@ -62,39 +62,49 @@ union WrapperData<T> {
 #[repr(C, align(8))]
 pub struct Wrapper<'a, Ta: 'a, Ts: 'static> {
     gc_info: u8,
+    data_offset: u8,
     data: WrapperData<Ta>,
     _phantom: PhantomData<&'a Ts>
 }
 
 impl<'a, Ta: 'a, Ts: 'static> Wrapper<'a, Ta, Ts> {
     pub fn owned(data: Ta) -> Self {
-        Self {
+        let mut r = Self {
             data: WrapperData {
                 value: ManuallyDrop::new(MaybeUninit::new(data))
             },
+            data_offset: 0,
             gc_info: GcInfo::Owned as u8,
             _phantom: PhantomData::default()
-        }
+        };
+        r.data_offset = (&r.data as *const _ as *const () as usize - &r as *const _ as *const () as usize) as u8;
+        r
     }
 
     pub fn shared(data: &Ta) -> Self {
-        Self {
+        let mut r = Self {
             data: WrapperData {
                 ptr: unsafe { NonNull::new_unchecked(data as *const Ta as *mut Ta) }
             },
+            data_offset: 0,
             gc_info: GcInfo::SharedFromHost as u8,
             _phantom: PhantomData::default()
-        }
+        };
+        r.data_offset = (&r.data as *const _ as *const () as usize - &r as *const _ as *const () as usize) as u8;
```

Comment:
This is necessary since user defined type does not always align to 8. We must measure the offset and store them somewhere.